content/Testing in Django - Basics - useful tools and good practices.md

```diff
@@ -0,0 +1,352 @@
+As a person that works in software development, I can say that it's a lot better when **you** discover an error rather than your client or a user in production. In this article, I'll talk about basic knowledge in automatic testing, useful tools, and good practices in [Django](https://www.djangoproject.com/) projects, with a focus on API's.
+
+## Testing in Django & Django REST Framework
+Django has a very nice [documentation](https://docs.djangoproject.com/en/3.1/topics/testing/) about testing, and [Django REST Framework](https://www.django-rest-framework.org/api-guide/testing/) too. So, in this blog, I'll talk about the main tools in those two frameworks, and what you can do and use to improve your testing.
+
+### Main testing classes
+Django provides several classes for testing. The one I'll talk about here is [TestCase](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.TestCase), very useful if your application uses databases.  
+As the name says, to create a test case in your Django project, you will define a class that inherits from TestCase. Doing this, you can use all the methods and properties of the named class that will help you to create and execute your tests. Then you will define different functions that will correspond to each unit test inside the test case. Let's see an example of the structure of a test case:
+
+```python
+# your test file
+from django.test import TestCase
+
+
+# class to define a test case for login
+class UserLoginTestCase(TestCase):
+
+    ...
+    # some setup here, explained later
+
+    def test_correct_login(self):
+        # unit test
+        # Corroborate the expected scenario
+        ...
+    
+    def test_if_password_incorrect_then_cant_login(self):
+        # unit test
+        # Corroborate that user's password needs to be only the correct one
+        ...
+    
+    def test_if_user_not_registered_cant_login(self):
+        # unit test
+        # Corroborate that user's are able to login only if they're registered
+    
+    ...
+```
+
+Also, you can use the [Client](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.Client) class of Django, that simulates a dummy browser, so you can make HTTP requests and test how your Django API responds.  
+
+Since this blog is focused in API testing, I want to talk about [APITestCase](https://www.django-rest-framework.org/api-guide/testing/#api-test-cases), a class of Django REST framework, that is a mirror of TestCase, but it uses a different client class: [APIClient](https://www.django-rest-framework.org/api-guide/testing/#apiclient). This client extends the TestClient. That means that has the same functionalities, and adds some other, such as the `credentials` function. This function is very useful to overwrite authentication headers, for example using OAuth1, OAuth2, or any simple token authentication scheme.  
```

Comment:
```suggestion
Since we are focusing on API testing, I want to talk about [APITestCase](https://www.django-rest-framework.org/api-guide/testing/#api-test-cases), a class of Django REST framework that is a mirror of TestCase but uses a different client class: [APIClient](https://www.django-rest-framework.org/api-guide/testing/#apiclient). This client extends the TestClient. Thhis means that it has the same functionalities and adds others such as the `credentials` function. This function is very useful to overwrite authentication headers, for example, using OAuth1, OAuth2, or any simple token authentication scheme.  
```